peps/pep-0810.rst

```diff
@@ -0,0 +1,1345 @@
+
+  def process(items: Sequence[str]) -> Mapping[str, int]:
+      ...
+
+Instead of:
+
+.. code-block:: python
+
+  from typing import TYPE_CHECKING
+  if TYPE_CHECKING:
+      from collections.abc import Sequence, Mapping
+
+  def process(items: Sequence[str]) -> Mapping[str, int]:
+      ...
+
+**Q: What's the performance overhead of lazy imports?**
+
+A: The overhead is minimal:
+
+- Zero overhead after first use thanks to the adaptive interpreter optimizing the slow path away.
+- Small one-time cost to create the proxy object.
+- Reification (first use) has the same cost as a regular import.
+- No ongoing performance penalty unlike ``importlib.util.LazyLoader``.
+
+Benchmarking with the `pyperformance suite`_ shows the implementation is performance
+neutral when lazy imports are not used.
+
+.. _pyperformance suite: https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20250922-3.15.0a0-27836e5/bm-20250922-vultr-x86_64-DinoV-lazy_imports-3.15.0a0-27836e5-vs-base.svg
+
+**Q: Can I mix lazy and eager imports of the same module?**
+
+A: Yes. If module ``foo`` is imported both lazily and eagerly in the same
+program, the eager import takes precedence and both bindings resolve to the same
+module object.
+
+**Q: How do I migrate existing code to use lazy imports?**
+
+A: Migration is incremental:
+
+1. Identify slow-loading modules using profiling tools
+2. Add ``lazy`` keyword to imports that aren't needed immediately
+3. Test that side-effect timing changes don't break functionality
+4. Use ``__lazy_modules__`` for compatibility with older Python versions
+
+**Q: What about star imports** (``from module import *``)?
+
+A: Wild card (star) imports cannot be lazy - they remain eager. This is because the
+set of names being imported cannot be determined without loading the module. Using the
+``lazy`` keyword with star imports will be a syntax error. If lazy imports are globally
+enabled, star imports will still be eager.
+
+**Q: How do lazy imports interact with import hooks and custom loaders?**
+
+A: Import hooks and loaders work normally. When a lazy object is first used, the
+standard import protocol runs, including any custom hooks or loaders that were
+in place at reification time.
+
+**Q: What happens in multi-threaded environments?**
+
+A: Lazy import reification is thread-safe. Only one thread will perform the
+actual import, and the binding is atomically updated. Other threads will see
+either the lazy proxy or the final resolved object.
+
+**Q: Can I force reification of a lazy import without using it?**
+
+A: Yes, accessing a module's ``__dict__`` will reify all lazy objects in that
+module. Individual lazy objects can be resolved by calling their ``get()`` method.
+
+**Q: What's the difference between** ``globals()`` **and** ``mod.__dict__`` **for lazy imports?**
+
+A: Calling ``globals()`` returns the module's dictionary without reifying lazy
```

Comment:
I guess this can be discussed on Discourse but I'm not sure I see why these two should be different.